core/embed/extmod/modtrezorcrypto/modtrezorcrypto-secp256k1_zkp.h

```diff
@@ -325,13 +331,265 @@ STATIC mp_obj_t mod_trezorcrypto_secp256k1_context_multiply(
+    mp_raise_ValueError("Generator parsing failed");
+  }
+
+  mp_buffer_info_t scratch_buffer;
+  mp_get_buffer_raise(args[8], &scratch_buffer, MP_BUFFER_WRITE);
+  if (scratch_buffer.len < RANGEPROOF_SIGN_BUFFER_SIZE) {
+    mp_raise_ValueError("Invalid length of output buffer");
+  }
+
+  size_t rangeproof_len = scratch_buffer.len;
+  if (!secp256k1_rangeproof_sign(
+          ctx, scratch_buffer.buf, &rangeproof_len, VALUE_MIN, &commitment,
+          blind.buf, nonce.buf, EXPONENT, BITS, value, message.buf, message.len,
+          extra_commit.buf, extra_commit.len, &generator)) {
+    mp_raise_ValueError("Rangeproof sign failed");
+  }
+  return mp_obj_new_memoryview('B', rangeproof_len, scratch_buffer.buf);
+}
+
+STATIC MP_DEFINE_CONST_FUN_OBJ_VAR_BETWEEN(
+    mod_trezorcrypto_secp256k1_context_rangeproof_sign_obj, 9, 9,
+    mod_trezorcrypto_secp256k1_context_rangeproof_sign);
+
+/// def rangeproof_rewind(self, conf_value: bytes, conf_asset: bytes,
+///                       nonce: bytes, range_proof: bytes,
+///                       extra_commit: bytes, asset_message: bytearray) ->
+///                       (value: long, blind: bytes,
+///                        asset_message: memoryview):
+///     '''
+///     Rewind a range proof to get the value, blinding factor and message.
+///     '''
+STATIC mp_obj_t mod_trezorcrypto_secp256k1_context_rangeproof_rewind(
+    size_t n_args, const mp_obj_t *args) {
+  const secp256k1_context *ctx =
+      mod_trezorcrypto_get_secp256k1_context(args[0]);
+
+  mp_buffer_info_t conf_value;
+  mp_get_buffer_raise(args[1], &conf_value, MP_BUFFER_READ);
+  secp256k1_pedersen_commitment commitment;
+  if (!secp256k1_pedersen_commitment_parse(ctx, &commitment, conf_value.buf)) {
+    mp_raise_ValueError("Invalid Pedersen commitment");
+  }
+
+  mp_buffer_info_t gen;
+  mp_get_buffer_raise(args[2], &gen, MP_BUFFER_READ);
+  if (gen.len != 33) {
+    mp_raise_ValueError("Invalid length of confidential asset");
+  }
+  secp256k1_generator generator;
+  if (!secp256k1_generator_parse(ctx, &generator, gen.buf)) {
+    mp_raise_ValueError("Generator parsing failed");
+  }
+
+  mp_buffer_info_t nonce;
+  mp_get_buffer_raise(args[3], &nonce, MP_BUFFER_READ);
+
+  mp_buffer_info_t range_proof;
+  mp_get_buffer_raise(args[4], &range_proof, MP_BUFFER_READ);
+
+  mp_buffer_info_t extra_commit;
+  mp_get_buffer_raise(args[5], &extra_commit, MP_BUFFER_READ);
+
+  mp_buffer_info_t asset_message;
+  mp_get_buffer_raise(args[6], &asset_message, MP_BUFFER_WRITE);
+
+  size_t asset_message_len = asset_message.len;
+  byte value_blind[32] = {0};
+  uint64_t value = 0;
+  uint64_t min_value = 0;
+  uint64_t max_value = 0;
+  if (!secp256k1_rangeproof_rewind(
+          ctx, value_blind, &value, asset_message.buf, &asset_message_len,
+          nonce.buf, &min_value, &max_value, &commitment, range_proof.buf,
+          range_proof.len, extra_commit.buf, extra_commit.len, &generator)) {
```

Comment:
This relies on the undocumented fact that `secp256k1_rangeproof_rewind` won't write more bytes than specified by it's 5th argument. https://github.com/ElementsProject/secp256k1-zkp/blob/secp256k1-zkp/src/modules/rangeproof/rangeproof_impl.h#L468
This is fine but we should add this guarantee to the upstream API (and maybe actually fail if the provided buffer was not large enough to hold the message instead of silently truncating the message without giving the caller a way to tell if truncation happened or not)

Reply:
> (and maybe actually fail if the provided buffer was not large enough to hold the message instead of silently truncating the message without giving the caller a way to tell if truncation happened or not)

Not sure if this part makes sense. Now that I look at it again, I think the message length cannot be determined by the rewinder (i.e., you cannot distinguish between a 96 bytes message and a 4096 byte message where the last 4000 bytes are 0x00). We'll need to look at this. 

Related: https://github.com/ElementsProject/secp256k1-zkp/issues/42

Reply:
We can re-use the pre-allocated `proof_buffer` (which is currently >5KB) for the message during the rewinding (currently it's only used for range/surjection proof generation).

Reply:
Unfortunately, since we need to hold both the user-provided rangeproof and the message buffer (for the `asset_message`) in memory in addition to the `secp256k1_context`, the unblinding start to fail with OOM errors.
In our case, we know that message should be 64 bytes - so IIUC it should be fine to "stop" the parsing (and truncate the rest of the message if it's longer), since we can re-blind the output to make sure we've got the correct asset and asset blind factor.

Reply:
>  the unblinding start to fail with OOM errors.

After a few memory optimizations, I think that we would be able fit the `proof_buffer` into RAM during unblinding :)
Will squash and rebase the PR with the fixes tomorrow.

Reply:
Done @ https://github.com/romanz/trezor-firmware/commit/1dd6e8972fb6854da1e9fdcedb08ef491d148330

Reply:
Update:
(Probably) due to heap fragmentation, large buffer allocation fail when ran after address/signature-related API invocations :(
I have added a compile flag at 7144d566 for allocating the largest buffer (containing the range-proof) statically at CCMRAM - which can be disabled when the stack would be moved there (https://github.com/trezor/trezor-firmware/issues/174).